light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+	var (
+		m   = make(map[uint64]int)
+		pos int
+	)
+	addToIndexMap(m, f, &pos, 1)
+	return m
+}
+
+// addToIndexMap recursively creates index associations for a given proof format subtree.
+func addToIndexMap(m map[uint64]int, f ProofFormat, pos *int, index uint64) {
+	l, r := f.children()
+	if l == nil {
+		m[index] = *pos
+		(*pos)++
+	} else {
+		addToIndexMap(m, l, pos, index*2)
+		addToIndexMap(m, r, pos, index*2+1)
+	}
+}
+
+// ChildIndex returns the generalized tree index of a subtree node in terms of
+// the main tree where a is the main tree index of the subtree root and b is the
+// subtree index of the node in question.
+func ChildIndex(a, b uint64) uint64 {
+	return (a-1)<<(63-bits.LeadingZeros64(b)) + b
+}
+
+// indexMapFormat implements ProofFormat based on an index map filled with
+// AddLeaf calls. Subtree formats can also be attached at certain indices.
+type indexMapFormat struct {
+	leaves map[uint64]ProofFormat
+	index  uint64
+}
+
+// NewIndexMapFormat returns an empty format.
+func NewIndexMapFormat() indexMapFormat {
+	return indexMapFormat{leaves: make(map[uint64]ProofFormat), index: 1}
+}
+
+// AddLeaf adds either a single leaf or attaches a subtree at the given tree index.
+func (f indexMapFormat) AddLeaf(index uint64, subtree ProofFormat) indexMapFormat {
+	if subtree != nil {
+		f.leaves[index] = subtree
+	}
+	for index > 1 {
+		index /= 2
+		f.leaves[index] = nil
+	}
+	return f
+}
+
+// children implements ProofFormat
+func (f indexMapFormat) children() (left, right ProofFormat) {
+	if st, ok := f.leaves[f.index]; ok {
+		if st != nil {
+			return st.children()
+		}
+		return indexMapFormat{leaves: f.leaves, index: f.index * 2}, indexMapFormat{leaves: f.leaves, index: f.index*2 + 1}
+	}
+	return nil, nil
+}
+
+// rangeFormat defined a proof format with a continuous range of leaf indices.
+// Attaching subtree formats is also possible.
+type rangeFormat struct {
+	begin, end, index uint64 // begin and end should be on the same level
+	subtree           func(uint64) ProofFormat
+}
+
+// NewRangeFormat creates a new rangeFormat with leafs in the begin..end range.
+// If subtrees != nil then further subtree formats are attached at indices where
+// the function returns a non-nil format.
+func NewRangeFormat(begin, end uint64, subtree func(uint64) ProofFormat) rangeFormat {
+	return rangeFormat{
+		begin:   begin,
+		end:     end,
+		index:   1,
+		subtree: subtree,
+	}
+}
+
+// children implements ProofFormat
+func (rf rangeFormat) children() (left, right ProofFormat) {
+	var (
+		lzr = bits.LeadingZeros64(rf.begin)
+		lzi = bits.LeadingZeros64(rf.index)
+	)
+	if lzi < lzr {
+		return nil, nil
+	}
+	if lzi == lzr {
+		if rf.subtree != nil && rf.index >= rf.begin && rf.index <= rf.end {
+			if st := rf.subtree(rf.index); st != nil {
+				return st.children()
+			}
+		}
+		return nil, nil
+	}
+	var (
+		i1 = rf.index << (lzi - lzr)
+		i2 = ((rf.index + 1) << (lzi - lzr)) - 1
+	)
+	if i1 <= rf.end && i2 >= rf.begin {
```

Comment:
For example, begin=8, and end=13; then i1=8, i2=15.

This condition is weird. `i1 <= rf.end && i2>=rf.begin` isn't always true unless index crosses this level where begin and end locates?

Reply:
It becomes false when no descendants of `rf.index` are in the `begin..end` range. In your example, for `rf.index == 7` it's false so index 7 will be stored in the proof but 14 and 15 won't be.
I added some explanatory commments to the code because it's probably not trivial at first glance to see what is happening.